### ISSUE TITLE
Upstream filters and filter state interaction.

### ISSUE THREAD

**Issue by paul-r-gall** [envoyproxy/envoy#24499]
*Title*: Crash on retry when writing read-only data from upstream filter.

*Description*:
Note: this has already been run by the envoy security team, and it's been decided that we can do this in the open.

Adding read-only data to downstream filter state in an upstream filter will crash Envoy when a retry happens. For example, the following test will crash envoy:

```
TEST_P(MultiplexedIntegrationTest, RetryWithFilterState) {
  test::integration::filters::HeaderToFilterStateFilterConfig config;
  config.set_header_name("x-uh-oh-header");
  config.set_state_name("uh-oh-state");
  config.set_read_only(true);
  envoy::extensions::filters::network::http_connection_manager::v3::HttpFilter filter;
  filter.set_name("header-to-filter-state");
  filter.mutable_typed_config()->PackFrom(config);
  config_helper_.prependFilter(MessageUtil::getYamlStringFromMessage(filter), false);
  // also add the x-uh-oh-header header.
  testRetry();
}
```

You might think that choosing to add the header in upstream filter state (`downstream_callbacks_->streamInfo().upstreamInfo()->upstreamStreamInfo()`) may be a better option, but it is not. This is because the upstream stream info is reset halfway through `UpstreamRequest::onPoolReady()`, which may be called after the filter chain has already decoded the headers.

*Call Stack*:
```
Unhandled exception:
    @     0x562c24625bb2  Envoy::StackTrace::terminateHandler()
    @     0x562c25539886  std::__terminate()
    @     0x562c25538f7b  __cxxabiv1::failed_throw()
    @     0x562c25538ef4  __cxa_throw
    @     0x562c2475f71f  Envoy::StreamInfo::FilterStateImpl::setData()
    @     0x562c22d0d0e0  Envoy::HeaderToFilterStateFilter::decodeHeaders()
    @     0x562c24273aa5  Envoy::Http::FilterManager::decodeHeaders()
    @     0x562c2424869d  Envoy::Router::Filter::doRetry()
    @     0x562c24bee4f7  event_process_active_single_queue
    @     0x562c24beab4c  event_base_loop
    @     0x562c22433721  Envoy::Server::WorkerImpl::threadRoutine()
    @     0x562c24bdd163  Envoy::Thread::ThreadImplPosix::ThreadImplPosix()::{lambda()#1}::__invoke()
    @     0x7f65400717d9  start_thread
    @     0x7f653ffe469f  clone
libc++abi: terminating due to uncaught exception of type Envoy::EnvoyException: FilterState::setData<T> called twice on same ReadOnly state.
```

**Comment by paul-r-gall** [comment]
cc @alyssawilk

**Comment by alyssawilk** [comment]
thanks for the report!  @paul-r-gall are you on this one or hoping I can pick it up?

**Comment by paul-r-gall** [comment]
I was planning on working on it, just cc'ing so you're in the loop.

**Comment by paul-r-gall** [comment]
Some more thoughts.

Currently there's (at least 3) unrelated filter states, accessible from upstream filters by the following methods:
1. `decoder_callbacks_->streamInfo().filterState()` This is downstream filter state, and ideally should not be mutable by upstream filters due to the retry exception thrown. Accessible from downstream access logs via "FILTER_STATE", but not accessible from upstream access logs.
2. `decoder_callbacks_->streamInfo().upstreamInfo()->upstreamFilterState()` This is shared state for the upstream, but has the afore mentioned footgun of not preserving writes from `decodeHeaders` due to a connection-level overwrite. This is accessible from both upstream and downstream access logs by "UPSTREAM_FILTER_STATE".
3. `decoder_callbacks_->upstreamCallbacks()->upstreamStreamInfo().filterState()`. This is also shared state for the upstream, but doesn't contain any content about the connection by default. There's no existing writes to this filter state in envoy code. This is accessible from upstream access logs by "FILTER_STATE", but not accessible from upstream access logs.

I would argue that filter states 2 and 3 should actually be the same filter state.

**Comment by kyessenov** [comment]
Sorry, I'm trying to understand the context for the PR #25917.
The crash is going to go away after https://github.com/envoyproxy/envoy/pull/25822. Is the underlying issue here that the upstream filters don't expect retries and sharing of the downstream filter state? Seems like a fundamental distinction and should be handled by the filters. This is not new - e.g. writing L4 filter state in L7 filter can crash sometimes because connections are shared.

As for the relationship between the filter states, I'd expect that upstream has the same structure of connection and request filter states. L7 streams are linked for HTTP and L4 streams are linked for TCP. A retry upstream should create a new filter state IMHO, to be consistent with downstream, and should use downstream filter state for something more persistent.

**Comment by paul-r-gall** [comment]
There's 2 underlying issues:
1. Upstream filters that write to downstream filter state in a read-only way will trigger the above ENVOY_BUG (after #25822) upon retries. If an upstream filter wants to write immutable L7 filter state, it should be writing to upstream filter state.
2. Upstream filter state creation ordering. The ordering is currently as follows:
  - Upstream Connection is attempted to be fetched (may be deferred)
  - Upstream Filter Chain decodeHeaders(..) is called
  - If deferred, Upstream Connection filter state *overwrites* whatever filter state the upstream filter chain added.

These two things combined mean that there's essentially no good way to write immutable filter state from decodeHeaders of the upstream filter chain. That's the issue that #25917 is attempting to address.
